obspy/core/tests/test_waveform_plugins.py

```diff
@@ -425,6 +426,47 @@ def test_deepcopy(self):
             msg = "Error in wavform format=%s" % format
             self.assertEqual(str(st), str(st_deepcopy), msg=msg)
 
+    def test_auto_file_format_during_writing(self):
+        """
+        The file format is either determined by directly specifying the
+        format or deduced from the filename. The former overwrites the latter.
+        """
+        # Get format name and name of the write function.
+        formats = [(key, value.module_name) for key, value in
+                   _get_default_eps('obspy.plugin.waveform',
+                                    'writeFormat').items()
+                   # Only test plugins that are actually part of ObsPy.
+                   if value.dist.key == "obspy"]
+
+        # Test for stream as well as for trace.
+        stream_trace = [read(), read()[0]]
+
+        for suffix, module_name in formats:
+            # A bit of magic to get the fully qualified function name...
+            fct_name = module_name + "." + load_entry_point(
+                "obspy",
+                "obspy.plugin.waveform.%s" % suffix,
+                "writeFormat").__name__
+            # For stream and trace.
+            for obj in stream_trace:
+                # Various versions of the suffix.
+                for s in [suffix.capitalize(), suffix.lower(), suffix.upper()]:
+                    with mock.patch(fct_name) as p:
+                        obj.write("temp." + s)
+                    # Make sure the fct has actually been called.
+                    self.assertEqual(p.call_count, 1)
+
+                    # Manually specifying the format name should overwrite
+                    # this.
+                    with mock.patch("obspy.io.mseed.core._write_mseed") as p:
+                        obj.write("temp." + s, format="mseed")
+                    self.assertEqual(p.call_count, 1)
+
+        # An unknown suffix should raise.
+        with self.assertRaises(TypeError):
```

Comment:
Euh, do we really raise a `TypeError`.. should be a `ValueError`. But has nothing to with this PR of course, just noticed right now.